wagtail/utils/utils.py

```diff
@@ -40,3 +44,49 @@ def flatten_choices(choices):
             # choice (key, display value)
             ret[str(key)] = value
     return ret
+
+
+def reduce_image_size(avatar, size_bound, max_dimensions=(400, 400)):
+    """
+    Reduce an image's file size to a size_bound in kilobytes.
+
+    :param avatar: The original image gotten from the image field of the form
+    :param size_bound: Desired size in KB
+    :param max_dimensions: Maximum dimensions for resizing
+    """
+    temp_buffer = BytesIO()
+    kilobyte = 1024
+    with Image.open(avatar) as img:
+        if img.mode == "RGBA":
+            img = img.convert("RGB")
+        img.thumbnail(max_dimensions, Image.LANCZOS)
+        img_ext = avatar.name.split(".")[-1]
+
+        quality = 95
+        step = 5
+
+        while quality > 10:
```

Comment:
I think this is probably overkill - if resizing an image to <400x400 doesn't immediately get it under 1Mb, then we've hit some kind of unusual edge case (maybe they've uploaded an enormous animated GIF, or the developer set `WAGTAIL_USER_PROFILE_IMAGE_SIZE_BOUND` to something absurdly low - which is another reason against making it a setting) and retrying it 20 times at decreasing quality is probably just going to waste CPU time.

As mentioned above, I don't think that hitting a specific threshold of 1Mb (or anything else) is particularly important to developers, as long as there's _some_ kind of guard against users uploading arbitrarily large images. I think the ideal implementation could be something as simple as:

* if either width or height is greater than 400, resize to 400x400
* otherwise, use the original file

...and ignore filesize entirely. Or, if you'd prefer to do an initial filesize check so that we're not making an up-front call to PIL for every image:

* if filesize is less than 1Mb, use the original file
* otherwise, resize to 400x400
* if the resulting file is smaller than the original, return the new one
* otherwise, return the original

Reply:
i guess i would ditch the default implementation and go with with 400x400. However, i only took this initial approach based on the issue this pull request addresses, so they can have a volition to prescribe a default they prefer but i think this suggestion is better and would go with it